sys/net/credman/credman.c

```diff
@@ -0,0 +1,167 @@
+#include "mutex.h"
+
+#include <string.h>
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+static mutex_t _mutex = MUTEX_INIT;
+
+static credman_credential_t credentials[CREDMAN_MAX_CREDENTIALS];
+static unsigned used = 0;
+
+static int _find_credential_pos(credman_tag_t tag, credman_type_t type);
+static int _find_next_free_pos(void);
+
+int credman_add(const credman_credential_t *credential)
+{
+    assert(credential);
+    mutex_lock(&_mutex);
+    int pos = -1;
+    int ret = CREDMAN_ERROR;
+
+    if ((credential->type == CREDMAN_TYPE_EMPTY) ||
+        (credential->tag == CREDMAN_TAG_EMPTY)) {
+        DEBUG("credman: invalid credential type/tag\n");
+        ret = CREDMAN_INVALID;
+        goto end;
+    }
+    switch (credential->type) {
+    case CREDMAN_TYPE_PSK:
+        if ((credential->params.psk.key.s == NULL) ||
+            (credential->params.psk.key.len == 0)) {
+            DEBUG("credman: invalid PSK parameters\n");
+            ret = CREDMAN_INVALID;
+            goto end;
+        }
+        break;
+    case CREDMAN_TYPE_ECDSA:
+        if ((credential->params.ecdsa.private_key == NULL) ||
+            (credential->params.ecdsa.public_key.x == NULL) ||
+            (credential->params.ecdsa.public_key.y == NULL)) {
+            DEBUG("credman: invalid ECDSA parameters\n");
+            ret = CREDMAN_INVALID;
+            goto end;
+        }
+        break;
+    default:
+        ret = CREDMAN_TYPE_UNKNOWN;
+        goto end;
+    }
+    pos = _find_credential_pos(credential->tag, credential->type);
+    if (pos >= 0) {
+        DEBUG("credman: credential with tag %d and type %d already exist\n",
+              credential->tag, credential->type);
+        ret = CREDMAN_EXIST;
+        goto end;
+    }
+    pos = _find_next_free_pos();
+    if (pos < 0) {
+        DEBUG("credman: no space for new credential\n");
+        ret = CREDMAN_NO_SPACE;
+        goto end;
+    }
+
+    memcpy(&credentials[pos], credential, sizeof(*credential));
+    used++;
+    ret = CREDMAN_OK;
+end:
+    mutex_unlock(&_mutex);
+    return ret;
+}
+
+int credman_get(credman_credential_t *credential,credman_tag_t tag,
+                credman_type_t type)
+{
+    assert(credential);
+    mutex_lock(&_mutex);
+    int ret = CREDMAN_ERROR;
+
+    int pos = _find_credential_pos(tag, type);
+    if (pos < 0) {
+        DEBUG("credman: credential with tag %d and type %d not found\n",
+               tag, type);
+        ret = CREDMAN_NOT_FOUND;
+        goto end;
+    }
+    memcpy(credential, &credentials[pos], sizeof(credman_credential_t));
+    ret = CREDMAN_OK;
+end:
+    mutex_unlock(&_mutex);
+    return ret;
+}
+
+int credman_delete(credman_tag_t tag, credman_type_t type)
+{
+    mutex_lock(&_mutex);
+    int ret = CREDMAN_ERROR;
+    int pos = _find_credential_pos(tag, type);
+    if (pos < 0) {
+        DEBUG("credman: credential with tag %d and type %d not found\n",
+               tag, type);
+        ret = CREDMAN_NOT_FOUND;
```

Comment:
IMHO it is okay to return `CREDMAN_OK` in this case

Reply:
So the whole function can be simplified to

```C
int credman_delete(credman_tag_t tag, credman_type_t type)
{
    mutex_lock(&_mutex);
    int ret = CREDMAN_ERROR;
    int pos = _find_credential_pos(tag, type);
    if (pos >= 0) {
        memset(&credentials[pos], 0, sizeof(credman_credential_t));
        used--;
    }
    mutex_unlock(&_mutex);
    return CREDMAN_OK;
}

Reply:
Going on that thought, I think it is better to also change the return value of `credman_delete` to `void` as returning `CREDMAN_OK` every time is not that useful after all. What's your opinion about this?

Reply:
:+1: